lib/dm-migrations/adapters/dm-do-adapter.rb

```diff
@@ -197,8 +197,14 @@ def create_unique_index_statements(model)
         def property_schema_hash(property)
           dump_class = property.dump_class
           type_map   = self.class.type_map
+          next_property_class = property.class
+          type_by_property_class = nil
+          while (type_by_property_class.nil? && next_property_class < DataMapper::Property) do
```

Comment:
I would also prefer a more explicit recursive approach (but I tried not to change the class signature with this commit). I'll update the pull request tomorrow (have to leave now, sorry).